### DISCUSSION TITLE
Testing of mtest's results

### DISCUSSION THREAD

**Discussion by endJunction** [thelfer/tfel#458] [Q&A] [answered]
Hi!

After developing an mfront model and writing mtests to verify its behaviour under different loads, how do you ensure the mtest results are reproducible over time?
The mtest has output of `.res` file, but I did not find an option in mtest to run comparison to the reference `.res`-file. Maybe there is already some mtest-test utility out there?

Few related thoughts:
 - `diff` is possible, but results might vary slightly and tolerances cannot be specified.
 - [`numdiff`]( https://www.nongnu.org/numdiff/) could be used, but it has no knowledge about the meaning of the columns or rows.
 - The precision of the data in the `.res` file might not be sufficient for strict testing---a binary format would be better.

 -- Best
  Dmitri

**Comment by thelfer** [comment, marked as answer]
Hi @endJunction,

You can have a look in the numerous tests in `mfront/tests/behaviours/generic`: https://github.com/thelfer/tfel/tree/master/mfront/tests/behaviours/generic

For example: https://github.com/thelfer/tfel/blob/master/mfront/tests/behaviours/generic/chaboche.mtest

You have the lines:

```cxx
@Test<file> @reference_file@ {'EXX':2,'EYY':3} 4.e-5 ; 
@Test<file> @reference_file@ {'SXX':8} 1.e0 ; 
```

You may want to have a look at `MFrontGallery` which provides built-in support for such tests. See https://github.com/thelfer/MFrontGallery/blob/master/materials/Concrete/behaviours/ and the file `ConcreteBurger_EDF_CIWAP_2021-test1.mtest`. The associated test is declared as follows  in the `CMakeLists.txt`:

```cmake
astermtest(Concrete ConcreteBurger_EDF_CIWAP_2021-test1
  MTEST_FILE ConcreteBurger_EDF_CIWAP_2021-test1
  LIBRARY $<TARGET_FILE:ConcreteBehaviours-aster>
  BEHAVIOUR asterconcreteburger_edf_ciwap_2021
  REFERENCE_FILE ${CMAKE_CURRENT_SOURCE_DIR}/ConcreteBurger_EDF_CIWAP_2021-test1.ref)
```

`astermtest` is a bit restrictive. See https://thelfer.github.io/MFrontGallery/web/cmake-infrastructure.html#functions-related-to-tests-based-on-mtest for more details

**Reply by endJunction** [reply]
Hi Thomas, this is great, thanks!

**Reply by thelfer** [reply]
Let me know if you need more help.

**Reply by endJunction** [reply]
Got it running for an mtest without substepping using both function and file for reference.

For a test which does substepping, the reference file test didn't work. It seems to me, that the comparison is executed for every substep but only full time steps are written into the `.res` file.

**Reply by thelfer** [reply]
The idea is reproductiblity. Even substepping must be reproduced.

**Reply by endJunction** [reply]
But how to get the substeps into the `.res` file? I didn't see a keyword among those for mtest....

**Reply by thelfer** [reply]
```bash
$ mtest --help-keyword=@OutputFrequency
The `OutputFrequency` let the user specify at when results are written
in the output files. This keyword is followed by a string. Two values
are allowed:

- `UserDefinedTimes`: the results are written at each time defined by
  the user. This is the default.
- `EveryIteration`: the outputs are written after each successful
  iteration.

This two options differs only in case of substepping.

## Example

~~~~ {.cpp}
@OutputFrequency 'EveryIteration';
~~~~
```

The name `EveryIteration` seems however unfortunate.

**Reply by thelfer** [reply]
Argh, the documentation was wrong. The right keyword is `EveryPeriod`.

**Reply by endJunction** [reply]
Ah! Yes, this fits now and I can proceed with testing. I appreciate the super fast replies!

**Reply by thelfer** [reply]
> Argh, the documentation was wrong. The right keyword is EveryPeriod.

https://github.com/thelfer/tfel/issues/459
